atc/api/artifactserver/create.go

```diff
@@ -0,0 +1,68 @@
+package artifactserver
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/concourse/baggageclaim"
+	"github.com/concourse/concourse/atc/api/present"
+	"github.com/concourse/concourse/atc/db"
+	"github.com/concourse/concourse/atc/worker"
+)
+
+func (s *Server) CreateArtifact(team db.Team) http.Handler {
+	hLog := s.logger.Session("create-artifact")
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+
+		// TODO: can probably check if fly sent us an etag header
+		// which we can lookup in the checksum field
+		// that way we don't have to create another volume.
+
+		spec := worker.VolumeSpec{
+			Strategy: baggageclaim.EmptyStrategy{},
+		}
+
+		volume, err := s.workerClient.CreateVolume(hLog, spec, team.ID(), db.VolumeTypeArtifact)
+		if err != nil {
+			hLog.Error("failed-to-create-volume", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		// NOTE: there's a race condition here between when the
+		// volume gets created and when the artifact gets initialized
+
+		// Within this timeframe there's a chance that the volume could
+		// get garbage collected out from under us.
+
+		// This happens because CreateVolume returns a 'created' instead
+		// of 'creating' volume.
+
+		// In the long run CreateVolume should probably return a 'creating'
+		// volume, but there are other changes needed in FindOrCreateContainer
+		// with the way we create volumes for a container
+
+		// I think leaving the race condition is fine for now. Worst case
+		// is a fly execute will fail and the user will need to rerun it.
+
+		artifact, err := volume.InitializeArtifact("", 0)
```

Comment:
'create' -> 'initialize' -> 'stream in data' feels unintuitive - normally it's 'initialized' only when it has all its data. it seems like this is to paper over the GC race condition. also why are both arguments zero-values?

Reply:
yes yes and yes. we can have this flow make more sense if we figure out the race condition.

lets just pretend those zero values aren't there... they get used for artifacts from build outputs.